File: December - 26/SakthiDharan01_mountain_climber_longest_ascending_path.java
import java.util.*;

public class SakthiDharan01_mountain_climber_longest_ascending_path {
    private static int m,n;
    private static int[][] grid;
    private static int[][] memo;
    private static final int[][] dirs = {{1,0},{-1,0},{0,1},{0,-1}};

    private static int dfs(int x,int y){
        if (memo[x][y]!=0) return memo[x][y];
        int best = 1;
        for (int[] d: dirs){
            int nx = x+d[0], ny = y+d[1];
            if (nx>=0 && nx<m && ny>=0 && ny<n && grid[nx][ny] > grid[x][y]){
                best = Math.max(best, 1 + dfs(nx, ny));
            }
        }
        memo[x][y] = best;
        return best;
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        m = sc.nextInt();
        if (!sc.hasNextInt()) return;
        n = sc.nextInt();
        grid = new int[m][n];
        memo = new int[m][n];
        for (int i=0;i<m;i++){
            for (int j=0;j<n;j++){
                if (sc.hasNextInt()) grid[i][j]=sc.nextInt();
            }
        }
        int ans = 0;
        for (int i=0;i<m;i++){
            for (int j=0;j<n;j++){
                ans = Math.max(ans, dfs(i,j));
            }
        }
        System.out.println(ans);
        sc.close();
    }
}

File: December - 26/SakthiDharan01_mountain_climber_longest_ascending_path.py
import sys
sys.setrecursionlimit(1_000_000)

def solve():
    data = list(map(int, sys.stdin.read().strip().split()))
    if len(data) < 2:
        return
    m, n = data[0], data[1]
    vals = data[2:2+m*n]
    grid = [vals[i*n:(i+1)*n] for i in range(m)]
    memo = [[0]*n for _ in range(m)]
    dirs = [(1,0),(-1,0),(0,1),(0,-1)]

    def dfs(x, y):
        if memo[x][y]:
            return memo[x][y]
        best = 1
        for dx, dy in dirs:
            nx, ny = x+dx, y+dy
            if 0<=nx<m and 0<=ny<n and grid[nx][ny] > grid[x][y]:
                best = max(best, 1 + dfs(nx, ny))
        memo[x][y] = best
        return best

    ans = 0
    for i in range(m):
        for j in range(n):
            ans = max(ans, dfs(i, j))
    print(ans)

if __name__ == "__main__":
    solve()
